BUG,DOC: Allow attach docs twice but error if wrong

seberg opened the pull request:
This stops general try/except, and instead skips setting of
docstrings if the docstring is identical. The latter part allows
for the import to be robust when e.g. reloading numpy which may
cause ``add_docstring`` to run twice.

This commit also changes the documentation stubs for scalar
attributes and errors to instead refer to the array attribute.
These are typically inherited, and thus it is not useful if
they refer to some "virtual attribute"

Closes gh-16209 and gh-14384

----

@rgommers to get this fixed for 1.19 quickly, had a look at the C-side fix, which unfportunately flushed out the exact thing I thought we might miss...

seberg commented on numpy/core/_add_newdocs.py:
```diff
@@ -5901,29 +5901,21 @@
+           refer_to_array_attribute('T', method=False))
+
+add_newdoc('numpy.core.numerictypes', 'generic',
+           refer_to_array_attribute('base', method=False))
```
For reviewers: If I had a typo in the attribute, this would crash. So the only thing is to not forget/double an attribute by accident.

seberg commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -1333,74 +1333,46 @@ gentype_transpose_get(PyObject *self)
     {"flat",
         (getter)gentype_flat_get,
-        (setter)0,
-        "a 1-d view of scalar",
```
Some of these may actually have been better docstrings in a sense. OTOH, refering the ndarray attribute is nice.

seberg commented on numpy/tests/test_reloading.py:
```diff
@@ -29,3 +31,30 @@ def test_novalue():
+    p = Process(target=try_full_reload)
+    p.start()
+    p.join()
+    assert p.exitcode == 0
```
Hmm, this fails on azure/windows with `FileNotFoundError: [Errno 2] No such file or directory: 'D:\\a\\1\\s\\build\\test\\runtests.py'` Just skip on windows:

https://dev.azure.com/numpy/numpy/_build/results?buildId=9919&view=logs&j=25bb66cf-4a16-533e-490c-8da4b5a3ec04&t=97446ee6-ebce-5662-24e5-f82eed312005&l=217

anirudh2290 replied:
other option is maybe try  https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods muliprocessing.set_start_method('spawn')

anirudh2290 replied:
not sure if this is already the default on windows for the version we are using

mattip replied:
Maybe 
```
txt = r"""
    import sys
    import numpy as np

    for k in list(sys.modules.keys()):
        if "numpy" in k:
            del sys.modules[k]

    import numpy as np
"""
p = subprocess.run([sys.executable, '-c', textwrap.dedent(txt)])
```

seberg replied:
Thanks matti, will just use that, seems simplest.

I changed the docs to link `ndarray.attribute`, did not realize the `~` would remove the ndarray as well.

seberg reviewed (commented):
And here the link to the artifact to see the actual doc changes quicker:

https://399-5652112-gh.circle-artifacts.com/0/doc/build/html/reference/generated/numpy.generic.html?highlight=generic#numpy.generic

rgommers commented:
> @rgommers to get this fixed for 1.19 quickly, had a look at the C-side fix, which unfportunately flushed out the exact thing I thought we might miss...

Thanks @seberg, happy for you to take this over - I'm pretty short on time on weekdays.

anirudh2290 commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -1482,7 +1482,7 @@ arr_add_docstring(PyObject *NPY_UNUSED(dummy), PyObject *args)
         if (!(doc)) {                                                   \
```
just a nit and unrelated to this change, but the do {...} while(0) can probably be removed. 
another nit and again unrelated to this change, but when i read this, i was searching for new in the code, but didnt realize that new was defined inside the macro, which made this a bit confusing.

seberg replied:
Yeah, lets move the new in front of the macro, also removes the need for the macro cast...

seberg replied:
I tried to clean it up, which deletes a bunch of other code since Python seems to define those types in the header now (if not, I assume CI will fail).

anirudh2290 reviewed (approved):
Overall LGTM ! 👍

eric-wieser commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -1507,7 +1507,8 @@ arr_add_docstring(PyObject *NPY_UNUSED(dummy), PyObject *args)
-        if (doc_attr != NULL && doc_attr != Py_None) {
+        if (doc_attr != NULL && doc_attr != Py_None &&
+                (PyUnicode_Compare(doc_attr, obj) != 0)) {
             PyErr_Format(PyExc_RuntimeError, "object %s", msg);
```
```suggestion
            if (PyErr_Occurred()) {
                return NULL;  // PyUnicode_Compare can fail
            }
            PyErr_Format(PyExc_RuntimeError, "object %s", msg);
```

seberg replied:
Right, seems there is also a %R or so missing to tell you which object we are talking about...

seberg replied:
Changed, the check was actually incorrect, it should have been `str` not `obj`, its fixed now (tested manually with `np.add_newdoc(np.take_along_axis, "wrong docstring")`)

BUT: We currently expose even `np.add_newdoc` as top-level... So I reverted the try/except change there, since it could be very disruptive in theory (i.e. something not importing). My tendency is to give a warning instead as soon as 1.20 is out.

eric-wieser commented on numpy/tests/test_reloading.py:
```diff
@@ -29,3 +33,25 @@ def test_novalue():
+
+        import numpy as np
+    """
+    p = subprocess.run([sys.executable, '-c', textwrap.dedent(code)])
```
Nit: I would always tend to decent at the same place as the literal

seberg committed:
TST: Add a test for np.add_docstring

Its not quite the right file, but close to newdoc seemed sensible
and we do not have a "right" file right now...

mattip commented:
You should skip the failing test on PyPy, you are running into gh-10167 where the call to `add_newdocs` should be replaced by generated C code that is compiled in and loaded before `PyType_Ready`

seberg commented:
Thanks, hadn't noticed the tests were failing...

mattip commented:
Now the `PYOPTIMIZE=2` CI run is failing the `TestAddDocstring.test_different_docstring_fail` tests.

seberg commented:
Dang, should have realized, passing now...

mattip commented:
The new tp_doc assigments seem to be working. [numpy.generice.base](https://412-5652112-gh.circle-artifacts.com/0/doc/build/html/reference/generated/numpy.generic.base.html) now looks something like this:


**numpy.generic.base**

attribute

generic.base:

Scalar attribute identical to the corresponding array attribute.

Please see [ndarray.base](https://412-5652112-gh.circle-artifacts.com/0/doc/build/html/reference/generated/numpy.ndarray.base.html#numpy.ndarray.base).

mattip commented:
LGTM

seberg commented:
Would be nice to finish this. @charris would  you prefer if I split this up to make the backport minimal?

charris commented:
Minimal backports are always nice :) I will probably do an rc2 next weekend as this appears a bit risky to me.

seberg committed:
BUG: Allow attaching documentation twice in add_docstring

This is technically not a bug, but some IDEs and IPython have
autoreload magic which can mean that NumPy gets reloaded a second
time. This is not safe, but when it happens ignoring that an
identical docstring is already attached fixes the issue.

seberg committed:
DOC: Fix generic scalar docstrings to point to ndarray

Previously some of these were not set, because they were already
set on the C-level. Remove the duplicate message and replace it
instead with a forward to the ``ndarray`` attribute/method.

These docstrings are inherited by the actual scalars and thus the
"virtual class" note was misleading.

seberg committed:
MAINT: Simplify logic in add_docstring

The old pointers are now provided by the C-API and the macros
seemed a bit confusing since ``new`` appears from nowhere being
defined in the macro.

seberg commented:
@charris I think it looks larger then it is and is only doc attaching, but still. The first commit is now the minimal fix. The `TST` commit is probably just as fine.

mattip commented:
Windows is complaining "ValueError: path is on mount 'D:', start on mount 'C:'", which was supposed to be avoided. Is this rebased off master?

seberg commented:
I did not rebase again, thought the commits apply easier then maybe, let me rebase...

seberg commented:
The windows builds seem super flaky today... but maybe time will make that go away...

mattip commented:
Close/Reopen to trigger CI merge with master

mattip commented:
LGTM. Can merge if CI passes

mattip commented on numpy/lib/tests/test_function_base.py:
```diff
@@ -3350,12 +3350,49 @@ class TestAdd_newdoc:
+    # Test should possibly be moved, but it also fits to be close to
+    # the newdoc tests...
+    @pytest.mark.skipif(sys.flags.optimize == 2, reason="Python running -OO")
+    def test_add_same_docstring(self):
```
This fails on PyPy since `np.ndarray.flat.__doc__` is None. Skipping would be the easiest fix I guess

seberg replied:
Thanks... dang, I had lost the fixups for this file in the rebase and forgot about this one...

mattip commented:
Thanks @seberg. ba14823f is the minimal change + test to be backported to fix this in 1.19, correct?
